File: ruby/lib/steam/community/steam_id.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'cgi'
require 'open-uri'
require 'rexml/document'

require 'exceptions/steam_condenser_exception'
require 'steam/community/cacheable'
require 'steam/community/game_stats'
require 'steam/community/steam_group'

# The SteamId class represents a Steam Community profile (also called Steam ID)
#
# @author Sebastian Staudt
class SteamId

  include Cacheable
  cacheable_with_ids :custom_url, :steam_id64

  # Returns the custom URL of this Steam ID
  #
  # The custom URL is a user specified unique string that can be used instead
  # of the 64bit SteamID as an identifier for a Steam ID.
  #
  # @note The custom URL is not necessarily the same as the user's nickname.
  # @return [String] The custom URL of this Steam ID
  attr_reader :custom_url

  # Returns the favorite game of this user
  #
  # @deprecated The favorite game is no longer listed for new users
  # @return [String] The favorite game of this user
  attr_reader :favorite_game

  # Returns the number of hours that this user played his/her favorite game in
  # the last two weeks
  #
  # @deprecated The favorite game is no longer listed for new users
  # @return [String] The number of hours the favorite game has been played
  #         recently
  attr_reader :favorite_game_hours_played

  # Returns the groups this user is a member of
  #
  # @return [Array<SteamGroup>] The groups this user is a member of
  attr_reader :groups

  # Returns the headline specified by the user
  #
  # @return [String] The headline specified by the user
  attr_reader :head_line

  # Returns the number of hours that this user played a game in the last two
  # weeks
  #
  # @return [Float] The number of hours the user has played recently
  attr_reader :hours_played

  # Returns the links that this user has added to his/her Steam ID
  #
  # The keys of the hash contain the titles of the links while the values
  # contain the corresponding URLs.
  #
  # @return [Hash<String, String>] The links of this user
  attr_reader :links

  # Returns the location of the user
  #
  # @return [String] The location of the user
  attr_reader :location

  # Returns the date of registration for the Steam account belonging to this
  # SteamID
  #
  # @return [Time] The date of the Steam account registration
  attr_reader :member_since

  # Returns the games this user has played the most in the last two weeks
  #
  # The keys of the hash contain the names of the games while the values
  # contain the number of hours the corresponding game has been played by the
  # user in the last two weeks.
  #
  # @return [Hash<String, Float>] The games this user has played the most
  #         recently
  attr_reader :most_played_games

  # Returns the Steam nickname of the user
  #
  # @return [String] The Steam nickname of the user
  attr_reader :nickname

  # Returns the privacy state of this Steam ID
  #
  # @return [String] The privacy state of this Steam ID
  attr_reader :privacy_state

  # Returns the real name of this user
  #
  # @return [String] The real name of this user
  attr_reader :real_name

  # Returns the message corresponding to this user's online state
  #
  # @return [String] The message corresponding to this user's online state
  # @see #ingame?
  # @see #online?
  attr_reader :state_message

  # Returns this user's 64bit SteamID
  #
  # @return [Fixnum] This user's 64bit SteamID
  attr_reader :steam_id64

  # Returns the Steam rating calculated over the last two weeks' activity
  #
  # @return [Float] The Steam rating of this user
  attr_reader :steam_rating

  # Returns the summary this user has provided
  #
  # @return [String] This user's summary
  attr_reader :summary

  # Returns the visibility state of this Steam ID
  #
  # @return [Fixnum] This Steam ID's visibility State
  attr_reader :visibility_state

  # Converts a 64bit numeric SteamID as used by the Steam Community to a
  # SteamID as reported by game servers
  #
  # @param [Fixnum] community_id The SteamID string as used by the Steam
  #        Community
  # @raise [SteamCondenserException] if the community ID is to small
  # @return [String] The converted SteamID, like `STEAM_0:0:12345`
  def self.convert_community_id_to_steam_id(community_id)
    steam_id1 = community_id % 2
    steam_id2 = community_id - 76561197960265728

    unless steam_id2 > 0
      raise SteamCondenserException, "SteamID #{community_id} is too small."
    end

    steam_id2 = (steam_id2 - steam_id1) / 2

    "STEAM_0:#{steam_id1}:#{steam_id2}"
  end

  # Converts a SteamID as reported by game servers to a 64bit numeric SteamID
  # as used by the Steam Community
  #
  # @param [String] steam_id The SteamID string as used on servers, like
  #        `STEAM_0:0:12345`
  # @raise [SteamCondenserException] if the SteamID doesn't have the correct
  #        format
  # @return [Fixnum] The converted 64bit numeric SteamID
  def self.convert_steam_id_to_community_id(steam_id)
    if steam_id == 'STEAM_ID_LAN' || steam_id == 'BOT'
      raise SteamCondenserException, "Cannot convert SteamID \"#{steam_id}\" to a community ID."
    elsif steam_id.match(/^STEAM_[0-1]:[0-1]:[0-9]+$/).nil?
      raise SteamCondenserException, "SteamID \"#{steam_id}\" doesn't have the correct format."
    end

    steam_id = steam_id[6..-1].split(':').map!{|s| s.to_i}

    steam_id[1] + steam_id[2] * 2 + 76561197960265728
  end

  # Creates a new `SteamId` instance using a SteamID as used on servers
  #
  # The SteamID from the server is converted into a 64bit numeric SteamID first
  # before this is used to retrieve the corresponding Steam Community profile.
  #
  # @param [String] steam_id The SteamID string as used on servers, like
  #        `STEAM_0:0:12345`
  # @return [SteamId] The `SteamId` belonging to the given SteamID
  # @see .convert_steam_id_to_community_id
  # @see #initialize
  def self.from_steam_id(steam_id)
    new(convert_steam_id_to_community_id(steam_id))
  end

  # Creates a new `SteamId` instance for the given Steam ID
  #
  # @param [String, Fixnum] id The custom URL of the Steam ID specified by the
  #        user or the 64bit SteamID
  # @param [Boolean] fetch if `true` the Steam ID's data is loaded into the
  #        object
  def initialize(id, fetch = true)
    begin
      if id.is_a? Numeric
        @steam_id64 = id
      else
        @custom_url = id.downcase
      end

      super(fetch)
    rescue REXML::ParseException
      raise SteamCondenserException, 'SteamID could not be loaded.'
    end
  end

  # Returns the base URL for this Steam ID
  #
  # This URL is different for Steam IDs having a custom URL.
  #
  # @return [String] The base URL for this SteamID
  def base_url
    if @custom_url.nil?
      "http://steamcommunity.com/profiles/#{@steam_id64}"
    else
      "http://steamcommunity.com/id/#{@custom_url}"
    end
  end

  # Fetchs data from the Steam Community by querying the XML version of the
  # profile specified by the ID of this Steam ID
  #
  # @raise SteamCondenserException if the Steam ID data is not available, e.g.
  #        when it is private
  # @see Cacheable#fetch
  def fetch
    profile_url = open(base_url + '?xml=1', {:proxy => true})
    profile = REXML::Document.new(profile_url.read).root

    unless REXML::XPath.first(profile, 'error').nil?
      raise SteamCondenserException, profile.elements['error'].text
    end

    @nickname         = CGI.unescapeHTML profile.elements['steamID'].text
    @steam_id64       = profile.elements['steamID64'].text.to_i
    @vac_banned       = (profile.elements['vacBanned'].text == 1)

    unless REXML::XPath.first(profile, 'privacyMessage').nil?
      raise SteamCondenserException, profile.elements['privacyMessage'].text
    end

    @image_url        = profile.elements['avatarIcon'].text[0..-5]
    @online_state     = profile.elements['onlineState'].text
    @privacy_state    = profile.elements['privacyState'].text
    @state_message    = profile.elements['stateMessage'].text
    @visibility_state = profile.elements['visibilityState'].text.to_i

    if @privacy_state == 'public'
      @custom_url                       = profile.elements['customURL'].text.downcase
      @custom_url                       = nil if @custom_url.empty?

      # The favorite game cannot be set since 10/10/2008, but old profiles
      # still have this. May be removed in a future version.
      unless REXML::XPath.first(profile, 'favoriteGame').nil?
        @favorite_game                  = profile.elements['favoriteGame/name'].text
        @favorite_game_hours_played     = profile.elements['favoriteGame/hoursPlayed2wk'].text
      end

      @head_line    = CGI.unescapeHTML profile.elements['headline'].text
      @hours_played = profile.elements['hoursPlayed2Wk'].text.to_f
      @location     = profile.elements['location'].text
      @member_since = Time.parse(profile.elements['memberSince'].text)
      @real_name    = CGI.unescapeHTML profile.elements['realname'].text
      @steam_rating = profile.elements['steamRating'].text.to_f
      @summary      = CGI.unescapeHTML profile.elements['summary'].text

      @most_played_games = {}
      unless REXML::XPath.first(profile, 'mostPlayedGames').nil?
        profile.elements.each('mostPlayedGames/mostPlayedGame') do |most_played_game|
          @most_played_games[most_played_game.elements['gameName'].text] = most_played_game.elements['hoursPlayed'].text.to_f
        end
      end

      @groups = []
      unless REXML::XPath.first(profile, 'groups').nil?
        profile.elements.each('groups/group') do |group|
          @groups << SteamGroup.new(group.elements['groupID64'].text.to_i, false)
        end
      end

      @links = {}
      unless REXML::XPath.first(profile, 'mostPlayedGames').nil?
        profile.elements.each('weblinks/weblink') do |link|
          @links[CGI.unescapeHTML link.elements['title'].text] = link.elements['link'].text
        end
      end
    end

    super
  end

  # Fetches the friends of this user
  #
  # This creates a new `SteamId` instance for each of the friends without
  # fetching their data.
  #
  # @see #friends
  # @see #initialize
  def fetch_friends
    url = "#{base_url}/friends?xml=1"

    @friends = []
    friends_data = REXML::Document.new(open(url, {:proxy => true}).read).root
    friends_data.elements.each('friends/friend') do |friend|
      @friends << SteamId.new(friend.text.to_i, false)
    end
  end

  # Fetches the games this user owns
  #
  # This fills the game hash with the names of the games as keys. The values
  # will either be `false` if the game does not have stats or the game's
  # "friendly name".
  #
  # @see #games
  def fetch_games
    url = "#{base_url}/games?xml=1"

    @games = {}
    games_data = REXML::Document.new(open(url, {:proxy => true}).read).root
    games_data.elements.each('games/game') do |game|
      game_name = game.elements['name'].text
      if game.elements['globalStatsLink'].nil?
        @games[game_name] = false
      else
        friendly_name = game.elements['globalStatsLink'].text.match(/http:\/\/steamcommunity.com\/stats\/([^?\/]+)\/achievements\//)[1]
        @games[game_name] = friendly_name.downcase
      end
    end

    true
  end

  # Returns the URL of the full-sized version of this user's avatar
  #
  # @return [String] The URL of the full-sized avatar
  def full_avatar_url
    "#{@image_url}_full.jpg"
  end

  # Returns the stats for the given game for the owner of this SteamID
  #
  # @param [String] game_name The friendly name of the game stats should be
  #        fetched for
  # @return [GameStats] The statistics for the game with the given name
  # @raise [ArgumentError] if the user does not own this game or it does not
  #        have any stats
  def game_stats(game_name)
    if games.has_value? game_name
      friendly_name = game_name
    elsif games.has_key? game_name.downcase
      friendly_name = games[game_name.downcase]
    else
      raise ArgumentError, "Stats for game #{game_name} do not exist."
    end

    GameStats.create_game_stats(@custom_url || @steam_id64, friendly_name)
  end

  # Returns the Steam Community friends of this user
  #
  # If the friends haven't been fetched yet, this is done now.
  #
  # @return [Array<SteamId>] The friends of this user
  # @see #fetch_friends
  def friends
    fetch_friends if @friends.nil?
    @friends
  end

  # Returns the games this user owns
  #
  # The keys of the hash are the games' names and the values are the "friendly
  # names" used for stats or `false` if the games has no stats.
  #
  # If the friends haven't been fetched yet, this is done now.
  #
  # @return [Hash<String, Object>] Pairs of game names and friendly names
  # @see #fetch_games
  def games
    fetch_games if @games.nil?
    @games
  end

  # Returns the URL of the icon version of this user's avatar
  #
  # @return [String] The URL of the icon-sized avatar
  def icon_url
    "#{@image_url}.jpg"
  end

  # Returns whether the owner of this SteamID is VAC banned
  #
  # @return [Boolean] `true` if the user has been banned by VAC
  def is_banned?
    @vac_banned
  end

  # Returns whether the owner of this SteamId is playing a game
  #
  # @return [Boolean] `true` if the user is in-game
  def is_in_game?
    @online_state == 'in-game'
  end

  # Returns whether the owner of this SteamID is currently logged into Steam
  #
  # @return [Boolean] `true` if the user is online
  def is_online?
    @online_state != 'offline'
  end

  # Returns the URL of the medium-sized version of this user's avatar
  #
  # @return [String] The URL of the medium-sized avatar
  def medium_avatar_url
    "#{@image_url}_medium.jpg"
  end

end

File: php/lib/steam/community/SteamId.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2011, Sebastian Staudt
 *
 * @license http://www.opensource.org/licenses/bsd-license.php New BSD License
 */

require_once STEAM_CONDENSER_PATH . 'exceptions/SteamCondenserException.php';
require_once STEAM_CONDENSER_PATH . 'steam/community/GameStats.php';
require_once STEAM_CONDENSER_PATH . 'steam/community/SteamGame.php';
require_once STEAM_CONDENSER_PATH . 'steam/community/SteamGroup.php';

/**
 * The SteamId class represents a Steam Community profile (also called Steam
 * ID)
 *
 * @author     Sebastian Staudt
 * @package    steam-condenser
 * @subpackage community
 */
class SteamId {

    /**
     * @var array
     */
    private static $steamIds = array();

    /**
     * @var string
     */
    private $customUrl;

    /**
     * @var int
     */
    private $fetchTime;

    /**
     * @var array
     */
    private $friends;

    /**
     * @var array
     */
    private $games;

    /**
     * @var string
     */
    private $nickname;

    /**
     * @var array
     */
    private $playtimes;

    /**
     * @var string
     */
    private $steamId64;

    /**
     * Returns whether the requested Steam ID is already cached
     *
     * @param string $id The custom URL of the Steam ID specified by the player
     *        or the 64bit SteamID
     * @return <var>true</var> if this Steam ID is already cached
     */
    public static function isCached($id) {
        return array_key_exists(strtolower($id), self::$steamIds);
    }

    /**
     * Clears the Steam ID cache
     */
    public static function clearCache() {
        self::$steamIds = array();
    }

    /**
     * Converts a 64bit numeric SteamID as used by the Steam Community to a
     * SteamID as reported by game servers
     *
     * @param string $communityId The SteamID string as used by the Steam
     *        Community
     * @return string The converted SteamID, like <var>STEAM_0:0:12345</var>
     * @throws SteamCondenserException if the community ID is to small
     */
    public static function convertCommunityIdToSteamId($communityId) {
        $steamId1  = substr($communityId, -1) % 2;
        $steamId2a = intval(substr($communityId, 0, 4)) - 7656;
        $steamId2b = substr($communityId, 4) - 1197960265728;
        $steamId2b = $steamId2b - $steamId1;

        if($steamId2a <= 0 && $steamId2b <= 0) {
            throw new SteamCondenserException("SteamID $communityId is too small.");
        }

        return "STEAM_0:$steamId1:" . (($steamId2a + $steamId2b) / 2);
    }

    /**
     * Converts a SteamID as reported by game servers to a 64bit numeric
     * SteamID as used by the Steam Community
     *
     * @param string $steamId The SteamID string as used on servers, like
     *        <var>STEAM_0:0:12345</var>
     * @return string The converted 64bit numeric SteamID
     * @throws SteamCondenserException if the SteamID doesn't have the correct
     *         format
     */
    public static function convertSteamIdToCommunityId($steamId) {
        if($steamId == 'STEAM_ID_LAN' || $steamId == 'BOT') {
            throw new SteamCondenserException("Cannot convert SteamID \"$steamId\" to a community ID.");
        }
        if(preg_match('/^STEAM_[0-1]:[0-1]:[0-9]+$/', $steamId) == 0) {
            throw new SteamCondenserException("SteamID \"$steamId\" doesn't have the correct format.");
        }

        $steamId = explode(':', substr($steamId, 6));
        $steamId = $steamId[1] + $steamId[2] * 2 + 1197960265728;

        return '7656' . $steamId;
    }

    /**
     * Creates a new <var>SteamID</var> instance or gets an existing one from
     * the cache for the profile with the given ID
     *
     * @param string $id The custom URL of the Steam ID specified by player or
     *        the 64bit SteamID
     * @param boolean $fetch if <var>true</var> the profile's data is loaded
     *        into the object
     * @param boolean $bypassCache If <var>true</var> an already cached
     *        instance for this Steam ID will be ignored and a new one will be
     *        created
     * @return SteamId The <var>SteamId</var> instance of the requested profile
     */
    public static function create($id, $fetch = true, $bypassCache = false) {
        $id = strtolower($id);
        if(self::isCached($id) && !$bypassCache) {
            $steamId = self::$steamIds[$id];
            if($fetch && !$steamId->isFetched()) {
                $steamId->fetchMembers();
            }

            return $steamId;
        } else {
            return new SteamId($id, $fetch);
        }
    }

    /**
     * Creates a new <var>SteamId</var> instance using a SteamID as used on
     * servers
     *
     * The SteamID from the server is converted into a 64bit numeric SteamID
     * first before this is used to retrieve the corresponding Steam Community
     * profile.
     *
     * @param string $steamId The SteamID string as used on servers, like
     *        <var>STEAM_0:0:12345</var>
     * @return SteamId The <var>SteamId</var> instance belonging to the given
     *         SteamID
     * @see convertSteamIdToCommunityId()
     * @see __construct()
     */
    public static function getFromSteamId($steamId) {
        return new SteamId(self::convertSteamIdToCommunityId($steamId));
    }

    /**
     * Creates a new <var>SteamId</var> instance for the given ID
     *
     * @param string $id The custom URL of the group specified by the player
     *        or the 64bit SteamID
     * @param boolean $fetch if <var>true</var> the profile's data is loaded
     *        into the object
     * @throws SteamCondenserException if the Steam ID data is not available,
     *         e.g. when it is private
     */
    public function __construct($id, $fetch = true) {
        if(is_numeric($id)) {
            $this->steamId64 = $id;
        } else {
            $this->customUrl = strtolower($id);
        }

        if($fetch) {
            $this->fetchData();
        }

        $this->cache();
    }

    /**
     * Saves this <var>SteamId</var> instance in the cache
     */
    public function cache() {
        if(!array_key_exists($this->steamId64, self::$steamIds)) {
            self::$steamIds[$this->steamId64] = $this;
            if(!empty($this->customUrl) &&
               !array_key_exists($this->customUrl, self::$steamIds)) {
               self::$steamIds[$this->customUrl] = $this;
            }
        }
    }

    /**
     * Fetchs data from the Steam Community by querying the XML version of the
     * profile specified by the ID of this Steam ID
     *
     * @throws SteamCondenserException if the Steam ID data is not available,
     *         e.g. when it is private, or when it cannot be parsed
     */
    public function fetchData() {
        $url = $this->getBaseUrl() . '?xml=1';
        $profile = new SimpleXMLElement(file_get_contents($url));

        if(!empty($profile->error)) {
            throw new SteamCondenserException((string) $profile->error);
        }

        $this->nickname  = htmlspecialchars_decode((string) $profile->steamID);
        $this->steamId64 = (string) $profile->steamID64;
        $this->vacBanned = (bool)(int) $profile->vacBanned;

        if(!empty($profile->privacyMessage)) {
            throw new SteamCondenserException((string) $profile->privacyMessage);
        }

        $this->imageUrl = substr((string) $profile->avatarIcon, 0, -4);
        $this->onlineState = (string) $profile->onlineState;
        $this->privacyState = (string) $profile->privacyState;
        $this->stateMessage = (string) $profile->stateMessage;
        $this->visibilityState = (int) $profile->visibilityState;

        if($this->privacyState == 'public') {
            $this->customUrl = strtolower((string) $profile->customURL);
            $this->favoriteGame = (string) $profile->favoriteGame->name;
            $this->favoriteGameHoursPlayed = (string) $profile->favoriteGame->hoursPlayed2wk;
            $this->headLine = htmlspecialchars_decode((string) $profile->headline);
            $this->hoursPlayed = (float) $profile->hoursPlayed2Wk;
            $this->location = (string) $profile->location;
            $this->memberSince = (string) $profile->memberSince;
            $this->realName = htmlspecialchars_decode((string) $profile->realname);
            $this->steamRating = (float) $profile->steamRating;
            $this->summary = htmlspecialchars_decode((string) $profile->summary);
        }

        if(!empty($profile->mostPlayedGames)) {
            foreach($profile->mostPlayedGames->mostPlayedGame as $mostPlayedGame) {
                $this->mostPlayedGames[(string) $mostPlayedGame->gameName] = (float) $mostPlayedGame->hoursPlayed;
            }
        }

        if(!empty($profile->groups)) {
            foreach($profile->groups->group as $group) {
                $this->groups[] = SteamGroup::create((string) $group->groupID64, false);
            }
        }

        if(!empty($profile->weblinks)) {
            foreach($profile->weblinks->weblink as $link) {
                $this->links[htmlspecialchars_decode((string) $link->title)] = (string) $link->link;
            }
        }

        $this->fetchTime = time();
    }

    /**
     * Fetches the friends of this user
     *
     * This creates a new <var>SteamId</var> instance for each of the friends
     * without fetching their data.
     *
     * @see getFriends()
     * @see __construct()
     * @throws SteamCondenserException if an error occurs while parsing the
     *         data
    */
    private function fetchFriends() {
        $url = $this->getBaseUrl() . '/friends?xml=1';

        $this->friends = array();
        $friendsData =  new SimpleXMLElement(file_get_contents($url));
        foreach($friendsData->friends->friend as $friend) {
            $this->friends[] = SteamId::create((string) $friend, false);
        }
    }

    /**
     * Fetches the games this user owns
     *
     * @see getGames()
     * @throws SteamCondenserException if an error occurs while parsing the
     *         data
     */
    private function fetchGames() {
        $this->games = array();
        $this->playtimes = array();

        $url = $this->getBaseUrl() . '/games?xml=1';
        $gamesData = new SimpleXMLElement(file_get_contents($url));

        foreach($gamesData->games->game as $gameData) {
            $game = SteamGame::create($gameData);
            $this->games[$game->getAppId()] = $game;
            $recent = (float) $gameData->hoursLast2Weeks;
            $total = (float) $gameData->hoursOnRecord;
            $playtimes = array((int) ($recent * 60), (int) ($total * 60));
            $this->playtimes[$game->getAppId()] = $playtimes;
        }
    }

    /**
     * Tries to find a game instance with the given application ID or full name
     * or short name
     *
     * @param mixed $id The full or short name or the application ID of the
     *        game
     * @return SteamGame The game found with the given ID
     * @throws SteamCondenserException if the user does not own the game or no
     *         game with the given ID exists
     */
    private function findGame($id) {
        $game = null;
        foreach($this->getGames() as $currentGame) {
            if($currentGame->getAppId() == $id ||
               $currentGame->getShortName() == $id ||
               $currentGame->getName() == $id) {
                $game = $currentGame;
                break;
            }
        }

        if($game == null) {
            if(is_int($id)) {
                $message = "This SteamID does not own a game with application ID {$id}.";
            } else {
                $message = "This SteamID does not own the game \"{$id}\".";
            }
            throw new SteamCondenserException($message);
        }

        return $game;
    }

    /**
     * Returns the base URL for this Steam ID
     *
     * This URL is different for Steam IDs having a custom URL.
     *
     * @return string The base URL for this SteamID
     */
    public function getBaseUrl() {
        if(empty($this->customUrl)) {
            return "http://steamcommunity.com/profiles/{$this->steamId64}";
        } else {
            return "http://steamcommunity.com/id/{$this->customUrl}";
        }
    }

    /**
     * Returns the time this group has been fetched
     *
     * @return int The timestamp of the last fetch time
     */
    public function getFetchTime() {
        return $this->fetchTime;
    }

    /**
     * Returns the Steam Community friends of this user
     *
     * If the friends haven't been fetched yet, this is done now.
     *
     * @return The friends of this user
     * @see fetchFriends()
     */
    public function getFriends() {
        if(empty($this->friends)) {
            $this->fetchFriends();
        }

        return $this->friends;
    }

    /**
     * Returns the URL of the full-sized version of this user's avatar
     *
     * @return string The URL of the full-sized avatar
     */
    public function getFullAvatarUrl() {
        return $this->imageUrl . '_full.jpg';
    }

    /**
     * Returns the games this user owns
     *
     * The keys of the hash are the games' application IDs and the values are
     * the corresponding game instances.
     *
     * If the friends haven't been fetched yet, this is done now.
     *
     * @return array The games this user owns
     * @see fetchGames()
     */
    public function getGames() {
        if(empty($this->games)) {
            $this->fetchGames();
        }

        return $this->games;
    }

    /**
     * Returns the stats for the given game for the owner of this SteamID
     *
     * @param mixed $id The full or short name or the application ID of the
     *        game stats should be fetched for
     * @return GameStats The statistics for the game with the given name
     * @throws SteamCondenserException if the user does not own this game or it
     *         does not have any stats
     */
    public function getGameStats($id) {
        $game = $this->findGame($id);

        if(!$game->hasStats()) {
            throw new SteamCondenserException("\"{$game->getName()}\" does not have stats.");
        }

        if(empty($this->customUrl)) {
            return GameStats::createGameStats($this->steamId64, $game->getShortName());
        } else {
            return GameStats::createGameStats($this->customUrl, $game->getShortName());
        }
    }

    /**
     * Returns the URL of the icon version of this user's avatar
     *
     * @return The URL of the icon-sized avatar
     */
    public function getIconAvatarUrl() {
        return $this->imageUrl . '.jpg';
    }

    /**
     * Returns the URL of the medium-sized version of this user's avatar
     *
     * @return The URL of the medium-sized avatar
     */
    public function getMediumAvatarUrl() {
        return $this->imageUrl . '_medium.jpg';
    }

    /**
     * Returns the Steam nickname of the user
     *
     * @return The Steam nickname of the user
     */
    public function getNickname() {
        return $this->nickname;
    }

    /**
     * Returns the time in minutes this user has played this game in the last
     * two weeks
     *
     * @param mixed $id The full or short name or the application ID of the
     *        game
     * @return int The number of minutes this user played the given game in the
     *         last two weeks
     */
    public function getRecentPlaytime($id) {
        $game = $this->findGame($id);
        $playtimes = $this->playtimes[$game->getAppId()];

        return $playtimes[0];
    }

    /**
     * Returns the total time in minutes this user has played this game
     *
     * @param mixed $id The full or short name or the application ID of the
     *        game
     * @return int The total number of minutes this user played the given game
     */
    public function getTotalPlaytime($id) {
        $game = $this->findGame($id);
        $playtimes = $this->playtimes[$game->getAppId()];

        return $playtimes[1];
    }

    /**
     * Returns whether the owner of this Steam ID is VAC banned
     *
     * @return <var>true</var> if the user has been banned by VAC
     */
    public function isBanned() {
        return $this->vacBanned;
    }

    /**
     * Returns whether the data for this Steam ID has already been fetched
     *
     * @return <var>true</var> if the Steam ID's data has been
     *         fetched
     */
    public function isFetched() {
        return !empty($this->fetchTime);
    }

    /**
     * Returns whether the owner of this Steam ID is playing a game
     *
     * @return <var>true</var> if the user is in-game
     */
    public function isInGame() {
        return $this->onlineState == 'in-game';
    }

    /**
     * Returns whether the owner of this Steam ID is currently logged into
     * Steam
     *
     * @return <var>true</var> if the user is online
     */
    public function isOnline() {
        return ($this->onlineState == 'online') || ($this->onlineState == 'in-game');
    }
}
?>
